design/one-pager-resource-connectivity-mvp.md

```diff
@@ -638,7 +638,279 @@ parameters:
+In addition, `RDSInstance`s also need following resources, so that they are accessible by the the worker nodes:
+
+* `AwsDbSubnetGroup`: represents a group of `AwsSubnet`s from different availability zones, from which a private IP is chosen and assigned to the RDS instance. 
+* `AwsSecurityGroup`: allows the RDS instance to accept traffic from a certain IP and port.
+
+To achieve resource connectivity goal in AWS, two approaches can be imagined:
+
+1. **Configuration based single managed controller:** Encapsulate all the external resource types in a general `AwsNetwork` type and create a single managed resource for it.
+2. **External resource based managed controller**: Create a corresponding managed resource for each required external resource type
+
+Although the first approach requires potentially less effort, the latter approach provides a few advantages:
+
+1. **Reusability**: these managed resource types could be later re-used for other configuration
+1. **YAML Configuration**: instead of implementing a specific configuration logic inside `AwsNetwork`, the configuration details are expressed in a YAML file. This makes creating more sophisticated configurations possible, without having to write controllers. 
+
+This document focuses on implementing the second approach.  For doing so, we address the following challenges:
+
+* As pointed out [earlier](#resource-references) in this document, each managed resource can refer to other resources either by their external names, or their object reference. In AWS case, since we are taking a YAML configuration approach, interconnecting resource by api objects makes a lot of sense. At the same time it is still useful to re-use existing external resources. Because of these requirements, we introduce the notion of *reference type* as following:
+
+	```yaml
+	...
+	spec:
+	  ...
+	  someParameter:
+	    # one of {external, apiObject} values
+	    referenceType: apiObject
+      # a string of attributeName.namespacedName format (apiObject case), or arbitrary (external case)
+	    value: attributeNAme.sampleObject.sampleNamespace
+	  ...
+	...
+	```
+
+* One of the issues that rises in this approach, is the dependency that a managed resource might take on another managed resource, when the former uses the latter in its spec. For example, in order to create a `AwsSubnet` instance which consumes a `AwsVpc` instance, the `AwsVpc` instance has to be created and ready. To solve this, we introduce the notion of *resource dependency* in consuming resource's spec:
+	```yaml
+	...
+	metadata:
+	  - namespace: example
+	    name: sample
+	...
+	spec:
+	  ...
+    dependsOn:
+      # a list of strings, each in namespacedName format
+  		- requiredResourceName.requiredResourceNamespace
+	  ...
+	...
+	```
+
+This indicates that `sample.example` resource needs to wait until all required resources are ready. Moreoveer, if a resource refers to another resource as an `apiObject` explained above, its dependency is implicit.
```

Comment:
I am not opposed to some sort of `dependsOn` functionality in Crossplane. However, we theoretically should be able to get by without it because of controllers should try to reconcile and create resources after a failure. Therefore, once the depended upon resource comes available, reconciliation should then succeed. It may be nice to not continually try if we know it is not going to work ahead of time though.

Reply:
That is true, and I agree that we could postpone adding this concept and just rely on try/error approach of eventual reconciliation. However, the nature of provisioning a resource is not always a transaction. As such trying provisioning a resource w/o having its resources in a ready state, might leave the resource in a state that it's provisioned, but it's now working properly (take dependency of eks cluster on cluster security group for instance)

Reply:
Opened #708 to address this separately